packages/tasit-action/src/TasitAction.test.js

```diff
@@ -113,126 +138,186 @@ describe("TasitAction.Contract", () => {
     });
 
-    it("should throw when subscribing with invalid trigger", async () => {
-      subscription = simpleStorage.setValue("hello world");
-      return expect(subscription.on("invalid", () => {})).to.be.rejected;
+    it("should throw when subscribing with invalid event name", async () => {
+      txSubscription = simpleStorage.setValue("hello world");
+
+      expect(() => {
+        txSubscription.on("invalid", () => {});
+      }).to.throw();
     });
 
     it("should throw when subscribing without listener", async () => {
-      subscription = simpleStorage.setValue("hello world");
-      return expect(subscription.on("confirmation")).to.be.rejected;
+      txSubscription = simpleStorage.setValue("hello world");
+
+      expect(() => {
+        txSubscription.on("confirmation");
+      }).to.throw();
+    });
+
+    it("should change contract state and trigger confirmation event one time", async () => {
+      txSubscription = simpleStorage.setValue(rand);
+      const confirmationFakeFn = sinon.fake();
+      const errorFakeFn = sinon.fake();
+
+      const confirmationListener = async message => {
+        const { data } = message;
+        const { confirmations } = data;
+
+        confirmationFakeFn();
+
+        const value = await simpleStorage.getValue();
+        expect(value).to.equal(rand);
+      };
+
+      txSubscription.once("confirmation", confirmationListener);
+
+      const errorListener = message => {
+        const { error, eventName } = message;
+        errorFakeFn();
+      };
+
+      txSubscription.on("error", errorListener);
+
+      await mineBlocks(provider, 15);
+
+      txSubscription.off("error");
+
+      expect(confirmationFakeFn.callCount).to.equal(1);
+      expect(errorFakeFn.called).to.be.false;
+      expect(txSubscription.eventNames()).to.be.empty;
     });
 
-    it("should call a write contract method (send tx) - immediate subscription", async () => {
-      var rand = Math.floor(Math.random() * Math.floor(1000)).toString();
-      subscription = simpleStorage.setValue(rand);
+    it("should change contract state and trigger confirmation event", async () => {
+      txSubscription = simpleStorage.setValue(rand);
       const fakeFn = sinon.fake();
 
-      const onMessage = async message => {
+      const listener = async message => {
         const { data } = message;
         const { confirmations } = data;
 
         if (confirmations >= 7) {
-          fakeFn();
-
-          subscription.removeAllListeners();
+          txSubscription.off("confirmation");
 
           const value = await simpleStorage.getValue();
           expect(value).to.equal(rand);
 
-          // FIXME: UnhandledPromiseRejectionWarning
-          //expect(1).to.equal(2);
+          fakeFn();
         }
       };
 
-      await subscription.on("confirmation", onMessage);
+      txSubscription.on("confirmation", listener);
 
       await mineBlocks(provider, 15);
```

Comment:
Yes, the removal of await keyword was made because `on()` is no longer an async function.

Reply:
Hmm sorry, my question was about the `mineBlocks` line

Reply:
Yes. With blocktime=0 no blocks are mined automatically by `ganache`, That should be made programmatically.

Reply:
Thanks for the info.

Unless a transaction is sent, in which case they are mined automatically, right?

But yes, I think for this many blocks it still makes sense to mine them.

Reply:
Makes sense, thanks!

Reply:
Note: Transaction events (our "confirmation" - ethers/web3 "hash","block") demands blocks mining. Contract events otherwise are triggered only waiting (promise) after tx broadcast.
I'm not sure on top of head why but I'll take a look.